cmd/juju/commands/upgradecontroller.go

```diff
@@ -64,41 +76,195 @@ func (c *upgradeControllerCommand) Info() *cmd.Info {
+	if c.modelConfigAPI != nil {
+		return c.modelConfigAPI, nil
 	}
-	return cmd.CheckEmpty(args)
+
+	api, err := c.NewModelAPIRoot(bootstrap.ControllerModelName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return modelconfig.NewClient(api), nil
+}
+
+func (c *upgradeControllerCommand) getControllerAPI() (controllerAPI, error) {
+	if c.controllerAPI != nil {
+		return c.controllerAPI, nil
+	}
+
+	return c.NewControllerAPIClient()
 }
 
 func (c *upgradeControllerCommand) Run(ctx *cmd.Context) (err error) {
-	// For now, we only support upgrading IAAS controllers.
-	// This is done by calling upgrade-juju -m controller.
 	controllerName, err := c.ControllerName()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if err := common.ValidateIaasController(c.CommandBase, c.Info().Name, controllerName, c.ClientStore()); err != nil {
+	controllerModel := jujuclient.JoinOwnerModelName(
+		names.NewUserTag(environs.AdminUser), bootstrap.ControllerModelName)
+	_, err = c.ModelUUIDs([]string{controllerModel})
+	if err != nil {
+		return errors.Annotatef(err, "cannot get controller model uuid")
+	}
+	details, err := c.ClientStore().ModelByName(controllerName, controllerModel)
+	if err != nil {
 		return errors.Trace(err)
 	}
+	if details.ModelType == model.CAAS {
+		return c.upgradeCAASController(ctx)
+	}
 	return c.upgradeIAASController(ctx)
 }
 
+func (c *upgradeControllerCommand) upgradeCAASController(ctx *cmd.Context) error {
+	if c.BuildAgent {
+		return errors.New("--build-agent for k8s controller upgrades")
+	}
+	client, err := c.getUpgradeJujuAPI()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	modelConfigClient, err := c.getModelConfigAPI()
+	if err != nil {
+		return err
+	}
+	defer modelConfigClient.Close()
+	controllerAPI, err := c.getControllerAPI()
+	if err != nil {
+		return err
+	}
+	defer controllerAPI.Close()
+
+	defer func() {
+		if err == errUpToDate {
+			ctx.Infof(err.Error())
+			err = nil
+		}
+	}()
+
+	// Determine the version to upgrade to.
+	attrs, err := modelConfigClient.ModelGet()
+	if err != nil {
+		return err
+	}
+	cfg, err := config.New(config.NoDefaults, attrs)
+	if err != nil {
+		return err
+	}
+
+	currentAgentVersion, ok := cfg.AgentVersion()
+	if !ok {
+		// Can't happen. In theory.
+		return errors.New("incomplete model configuration")
+	}
+
+	warnCompat, err := c.precheck(ctx, currentAgentVersion)
+	if err != nil {
+		return err
+	}
+
+	controllerCfg, err := controllerAPI.ControllerConfig()
+	if err != nil {
+		return err
+	}
+
+	context, err := c.initCAASVersions(controllerCfg, currentAgentVersion, warnCompat)
+	if err != nil {
+		return err
+	}
+
+	if err := context.maybeChoosePackagedAgent(); err != nil {
+		ctx.Verbosef("%v", err)
+		return err
+	}
+
+	if err := context.validate(); err != nil {
+		return err
+	}
+	ctx.Verbosef("available agent images:\n%s", formatVersions(context.packagedAgents))
+	fmt.Fprintf(ctx.Stderr, "best version:\n    %v\n", context.chosen)
```

Comment:
do we want to print to stderr here?

Reply:
this is what is done in the upgrade-model command so I wanted to keep it the same here
i am not sure why it was done like that